### ISSUE TITLE
Investigate restoring of cur_L on exceptional path rationale

### ISSUE THREAD

**Issue by Buristan** [tarantool/tarantool#6189]
The patches [`ed412cd9`](https://github.com/tarantool/luajit/commit/ed412cd9f55fe87fd32a69c86e1732690fc5c1b0) and [`97699d9e`](https://github.com/tarantool/luajit/commit/97699d9ee2467389b6aea21a098e38aff3469b5f) declare that without `cur_L` restoration of exceptional Tarantool is crashed due to `L->base` inconsistency with `cur_L` usage in `lj_vm_exit_handler()`. Unfortunately, the example mentioned in the corresponding issue (#1516) is no more valid.

Indeed, the inconsistency is occurring (with reverting patches) in some tests with spontaneously SegFault.

The minimum available flaky (10% of failures) reproducer for now:
```lua
box.cfg {
    listen              = "/tmp/vardir/tmp/box.socket-iproto",
    memtx_memory        = 107374182,
    pid_file            = "tarantool.pid",
}

local net_box = require('net.box')

local LISTEN = {
  host = "unix/",
  service = "/tmp/vardir/tmp/box.socket-iproto",
}

function fret1(arg1) local res = arg1 + #_G return res end
box.schema.user.grant('guest','read,write,execute','universe', nil, {if_not_exists=true})

print(#_G)

for i=1,100 do
        local cn = net_box.connect(LISTEN.host, LISTEN.service)
        cn:eval("fret1(#_G)")
        cn:close()
end
print("AFTER EVALS")

os.exit()
```

In the scope of this issue the stable test case (as simple as possible -- use module `fiber` and no others) should be implemented.

**Comment by Buristan** [comment]
**TLDR**: The reason of core dumps is Lua/C API misuse. An error is thrown on non-currently executed coroutine.

About the example in the issue header.

`tarantool_L` (`cur_L`) has the following Lua backtrace at the moment of error throwing on the non-currently executed coroutine (`L`):

```
#1: wait in =[C] at line -1
#2: wait_state in @builtin/box/net_box.lua at line 464
#3: connect in @builtin/box/net_box.lua at line 1068
#4: (unnamed) in @/tmp/former_schema.lua at line 25
```

Error is thrown on the other `L` with the folowing Lua backtrace:
```
#1: send_and_recv in =[C] at line -1
#2: send_and_recv_iproto in @builtin/box/net_box.lua at line 704
#3: (unnamed) in @builtin/box/net_box.lua at line 872
```

The main reason is the parameter `L` passed throw the `luaL_testcancel()` function, which is not the same as `fiber()->storage.L` (`cur_L`) -- currently executed coroutine.

After error catching by `pcall()` or `xpcall()` the `cur_L` state equals to old executed coroutine (from which error was thrown), so if after it `lj_vm_exit_handler()` is called (for example, after snapshot restoration) it restores base of the Lua stack from the wrong coroutine, that leads to stack inconsistency and crash.

See also LuaJIT/LuaJIT#740.

Currently, the LuaJIT will be adjusted for all architectures by the same patch as in the aforementioned pull request. In the Q3 `luaL_testcancel()` behaviour should be fixed on Tarantool side, and the assert should be added instead the fix.

**Comment by igormunkin** [comment]
The issue is resolved in scope of tarantool/luajit@7570ff6. The submodule was updated in tarantool in 2.9.0-345-g847b096e3 (847b096).
